### PR TITLE
Megatron Tensorizer Integration

### PR DESCRIPTION
**Pull request by harubaru**
This PR introduces support for [Tensorizer](https://github.com/coreweave/tensorizer) which allows for fast(er) checkpoint saving and loading. The motivation behind this is to speed up the process of saving and loading model checkpoints.

- Tensorizer can be activated in the training process by supplying the `--use-tensorizer` flag.
- The model weight initialization had been rewritten in this PR as during loading, Megatron initialized model weights on GPU which used an incredible amount of resources. To fix this, Tensorizer's `no_init_or_tensor` function was used to initialize the model on the `meta` device to save resources during checkpoint loading.
- Benchmarking tools and scripts are provided from within this PR as well as the `--timing-file` flag which is used to record checkpoint saving and loading times to a file.
- The changes work by serializing and deserializing individual components of the model such as the model weights, optimizer weights, and gradient weights. The PR was also only tested on decoder-only transformer models so it may not work for other types of transformer models such as encoder-decoder transformers.
- Information that is necessary for recreating the training state is still saved as a PyTorch pickle. It might be interesting to investigate storing such information through JSON but the vast bulk of the loading/saving time is instead spent on weights related to training the model.

### CODE REVIEW

File: megatron/training.py [hunk 1]
```diff
@@ -669,12 +723,27 @@ def training_log(loss_dict, total_loss_dict, learning_rate, iteration,
...
     # all ranks report the max time.
     timers('save-checkpoint', log_level=0).start(barrier=True)
-    save_checkpoint(iteration, model, optimizer, opt_param_scheduler)
+    if not args.use_tensorizer:
+        save_checkpoint(iteration, model, optimizer, opt_param_scheduler)
+    else:
+        save_checkpoint_tensorizer(iteration, model, optimizer, opt_param_scheduler)
     timers('save-checkpoint').stop(barrier=True)
-    timers.log(['save-checkpoint'])
+    if args.timing_file is not None:
```
Comment (wbrown) [inline megatron/training.py]: This looks like repeated copy-and-pasted code that can be converted into a function. The only diference is the `load_checkpoint` and `save_checkpoint` keys.

File: megatron/checkpointing.py [hunk 2]
```diff
@@ -211,6 +215,388 @@ def get_rng_state():
...
+    flattened, skeleton = flatten_dict_to_skeleton(opt.state_dict())
+    serializer = TensorSerializer(f'{checkpoint_name}-opt.tensors')
+    serializer.write_state_dict(flattened)
+    serializer.close()
+    json.dump(skeleton, fp=open(f'{checkpoint_name}-opt.json', 'w'))
+
+
+def load_optimizer(checkpoint_name):
+    opt_state_dict = json.load(fp=open(f'{checkpoint_name}-opt.json', 'r'))
+    deserializer = TensorDeserializer(f'{checkpoint_name}-opt.tensors', device=torch.cuda.current_device(), plaid_mode=True)
```
Comment (wbrown) [inline megatron/checkpointing.py]: What are the column length standards for the megatron repo?

File: megatron/checkpointing.py [hunk 3]
```diff
@@ -211,6 +215,388 @@ def get_rng_state():
...
+            elif isinstance(d, dict):
+                d = d.setdefault(key, {})
+        if isinstance(d, list):
+            d[int(keys[-1])] = v
+        else:
+            d[keys[-1]] = v
+    return skel
+
+
+def convert_parameters_to_tensors(d):
```
Comment (wbrown) [inline megatron/checkpointing.py]: Ugly. Is this mutating in place?

File: megatron/checkpointing.py [hunk 4]
```diff
@@ -211,6 +215,388 @@ def get_rng_state():
...
+        state_dict = torch.load(checkpoint_name, map_location='cpu')
+    except BaseException as e:
+        print_rank_0('could not load the tensorizer checkpoint')
+        print_rank_0(e)
+        sys.exit()
+
+    return state_dict, release, checkpoint_name
+
+
+def flatten_dict_to_skeleton(d, parent_key=''):
```
Comment (wbrown) [inline megatron/checkpointing.py]: Type annotate the return value please.

File: megatron/checkpointing.py [hunk 5]
```diff
@@ -211,6 +215,388 @@ def get_rng_state():
...
+                np.random.set_state(rng_state['np_rng_state'])
+                torch.set_rng_state(rng_state['torch_rng_state'])
+                torch.cuda.set_rng_state(rng_state['cuda_rng_state'])
+                # Check for empty states array
+                if not rng_state['rng_tracker_states']:
+                    raise KeyError
+                tensor_parallel.get_cuda_rng_tracker().set_states(
+                    rng_state['rng_tracker_states'])
+            else:  # backward compatability
+                random.setstate(state_dict['random_rng_state'])
```
Comment (wbrown) [inline megatron/checkpointing.py]: Can this be DRY'ed up?

**Review by wbrown — CHANGES_REQUESTED** [review]

File: megatron/checkpointing.py (hunk 2 above)
Comment (harubaru) [reply megatron/checkpointing.py]: It doesn't seem to be explicitly stated but the authors seem to style it consistently with [PEP 8 style guide](https://peps.python.org/pep-0008/#maximum-line-length) or max length of 79 characters. The original authors also use a [PEP 8 linter](https://github.com/NVIDIA/Megatron-LM/blob/fab0bd693ec5be55b32c4f12a1ea44766ec63448/tools/linter.py#L31).
